### PR TITLE
Multiple custom layouts for tracing views

### LINKED ISSUE #3230: Pre-defined and/or customizable layout presets
* [x] Add UI support for all kinds of tracing views
* [x] create a schema to save all different layouts in local storage
* [x] implement the schema and functionality to add and manage all layouts

### PR DESCRIPTION
**Pull request by MichaelBuessemeyer**
This PR addes the possibility to add muliple layouts.

### URL of deployed dev instance (used for testing):
- https://mulitpletracingcustomlayouts.webknossos.xyz/

### Steps to test:
- open a tracing
- create a new layout (should have the previous layout set as initial layout)
- change freely between the layouts
- try deleting a non-active layout
- an active layout should not be removable -> thus guaranteeing to always have a layout 
- the last active layout at closing a tracing, should be the initial layout on next startup
- important -> try this in all four kinds of tracing views as their layout is treated separately
- test the migration into the new "layout saving schema"

### Issues:
- fixes #3230

------
- [x] Updated [changelog](../blob/master/CHANGELOG.md#unreleased)
- ~~~[ ] Updated [migration guide](../blob/master/MIGRATIONS.md#unreleased) if applicable~~~
- ~~~[ ] Updated [documentation](../blob/master/docs) if applicable~~~
- ~~~[ ] Needs datastore update after deployment~~~
- [x] Ready for review

### CODE REVIEW

**Comment by MichaelBuessemeyer** [comment]
@philippotto shall we / I add a short "feature explanation" to the documentation?  Right now I can't find anything about the custom viewports / layout feature.

File: app/assets/javascripts/oxalis/view/layouting/layout_persistence.js [hunk 1]
```diff
@@ -18,56 +24,137 @@ const localStorageKeys = {
...
+    persistLayoutConfigsDebounced();
+    return true;
+  }
+  Toast.warning("Could not create a new Layout.");
+  return false;
+}
+
+export function setActiveLayout(layoutKey: LayoutKeys, activeLayout: string) {
+  const newLayouts = getDeepCopyOfStoredLayouts();
+  if (newLayouts[layoutKey] && newLayouts[layoutKey][activeLayout]) {
```
Comment (philippotto) [inline app/assets/javascripts/oxalis/view/layouting/layout_persistence.js]: Shouldn't we throw an error if this if-check fails?

File: package.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "webknossos",
-  "version": "0.17.146",
+  "version": "1.10.1",
```
Comment (philippotto) [inline package.json]: I think this change was added when you executed `yarn version`. Can you undo this?

File: app/assets/javascripts/oxalis/view/layouting/layout_persistence.js [hunk 3]
```diff
@@ -18,56 +24,137 @@ const localStorageKeys = {
...
+    return defaultLayoutSchema;
+  }
+  try {
+    const version = JSON.parse(storedLayoutVersion);
+    const layouts = JSON.parse(layoutString);
+    if (currentLayoutVersion > version) {
+      if (version !== 5) {
+        return defaultLayoutSchema;
+      }
+      // migrate to newst schema
```
Comment (philippotto) [inline app/assets/javascripts/oxalis/view/layouting/layout_persistence.js]: newEst

File: app/assets/javascripts/oxalis/view/layouting/layout_persistence.js [hunk 4]
```diff
@@ -18,56 +24,137 @@ const localStorageKeys = {
...
+          "Custom Layout": layouts.VolumeTracingView || defaultLayouts.VolumeTracingView,
+          lastActive: "Custom Layout",
+        },
+        ArbitraryLayout: {
+          "Custom Layout": layouts.ArbitraryLayout || defaultLayouts.ArbitraryLayout,
+          lastActive: "Custom Layout",
+        },
+        OrthoLayout: {
+          "Custom Layout": layouts.OrthoLayout || defaultLayouts.OrthoLayout,
+          lastActive: "Custom Layout",
```
Comment (philippotto) [inline app/assets/javascripts/oxalis/view/layouting/layout_persistence.js]: This `lastActive` attribute has some disadvantages in my opinion. Two example symptoms are that a layout must not be named "lastActive" and that "lastActive" has to be ignored when listing layout names. In general, the problem is that OrthoLayout (and co) contain a map from layout name to layout config AND the information what is active. These are two different things which can be stored independently.
For example, you could have:

```
...
OrthoLayout: {...},
LastActiveLayouts: {
  OrthoLayout: "Custom Layout",
  ArbitraryLayout: "Layout X",
}
```

as a format. Do you agree?

Comment (MichaelBuessemeyer) [reply app/assets/javascripts/oxalis/view/layouting/layout_persistence.js]: Sure it was obvious to me that my solution is not so good. I like your idea way more than mine and it makes the code even better :smiley:

File: app/assets/javascripts/oxalis/view/layouting/layout_persistence.js (hunk 1 above)
Comment (MichaelBuessemeyer) [reply app/assets/javascripts/oxalis/view/layouting/layout_persistence.js]: Like an error toast that says e.g. "The selected layout was not found. Please add a layout for the selected name."?

File: app/assets/javascripts/oxalis/view/action-bar/tracing_actions_view.js [hunk 5]
```diff
@@ -32,27 +31,109 @@ type StateProps = {
...
+    onSelectLayout,
+    onDeleteLayout,
+    addNewLayout,
+    ...others
+  } = props;
+  const layoutMissingHelpTitle = (
+    <React.Fragment>
+      <h5 style={{ color: "#fff" }}>Where is my layout?</h5>
+      <p>
+        {`The tracing views are seperated into four classes. Each of them has their own layouts. If you
```
Comment (philippotto) [inline app/assets/javascripts/oxalis/view/action-bar/tracing_actions_view.js]: `separated`

File: app/assets/javascripts/oxalis/view/layouting/default_layout_configs.js [hunk 6]
```diff
@@ -87,5 +106,12 @@ export function determineLayout(controlMode: ControlMode, viewMode: Mode): Layou
   }
 }
 
+export const mapLayoutKeysToLanguage = {
+  OrthoLayoutView: "Othogonal Tracings - View Only",
+  VolumeTracingView: "Volume Tracings",
+  ArbitraryLayout: "Arbitray View Mode",
+  OrthoLayout: "Othogonal Tracings",
```
Comment (philippotto) [inline app/assets/javascripts/oxalis/view/layouting/default_layout_configs.js]: oRthogonal (same in line 110)

Comment (philippotto) [reply app/assets/javascripts/oxalis/view/layouting/default_layout_configs.js]: Can you rename "Tracings" to "Mode" in these labels?

**Comment by philippotto** [comment]
Nice, nice, nice :) I made another comment ("Tracing" vs "Mode") and I also have another request: Can we remove the bullet points in front of the custom layout names? I think it would look better without the bullets.

**Comment by MichaelBuessemeyer** [comment]
@philippotto could you please have a look at this PR again? I applied the feedback, but the frontend test did run correctly the first time on CircleCI. But without changing a thing the second try worked out. I don't know why :thinking:

**Comment by philippotto** [comment]
> @philippotto could you please have a look at this PR again? I applied the feedback, but the frontend test did run correctly the first time on CircleCI. But without changing a thing the second try worked out. I don't know why thinking

So, both runs worked out? :) 

The PR looks really good. I also just tested the conversion between the persisted layouts and it seems to work pretty well.

What's a bit weird is that I can sometimes delete the last layout (so that the list is empty) and sometimes I can't delete it. But I think that's not a blocker. Feel free to merge!

**Comment by MichaelBuessemeyer** [comment]
>  but the frontend test did run correctly the first time on CircleCI.

 meant: "but the frontend test didn't run correctly the first time on CircleCI." :slightly_smiling_face:

**Pull request merged**